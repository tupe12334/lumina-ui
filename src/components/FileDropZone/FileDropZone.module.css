/* FileDropZone Container */
.container {
  width: 100%;
  max-width: 600px;
  margin: 0 auto;
}

/* Drop Zone */
.dropzone {
  position: relative;
  min-height: 200px;
  padding: 24px;
  border: 2px dashed #d1d5db;
  border-radius: 12px;
  background: linear-gradient(135deg, #fafafa 0%, #f9fafb 100%);
  transition: all 0.4s cubic-bezier(0.4, 0, 0.2, 1);
  cursor: pointer;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  text-align: center;
  outline: none;
}

.dropzone:hover {
  border-color: #9ca3af;
  background: linear-gradient(135deg, #f3f4f6 0%, #f1f5f9 100%);
  transform: translateY(-1px);
  box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1);
}

.dropzone:focus-visible {
  border-color: #3b82f6;
  box-shadow: 0 0 0 3px rgba(59, 130, 246, 0.1);
}

.dropzone.dragActive {
  border-color: #3b82f6;
  background: linear-gradient(135deg, rgba(59, 130, 246, 0.1) 0%, rgba(147, 197, 253, 0.1) 100%);
  transform: scale(1.02) translateY(-2px);
  box-shadow: 0 16px 25px -5px rgba(59, 130, 246, 0.25);
}

.dropzone.disabled {
  cursor: not-allowed;
  opacity: 0.6;
  border-color: #e5e7eb;
  background: #f9fafb;
}

.dropzone.disabled:hover {
  transform: none;
  box-shadow: none;
  border-color: #e5e7eb;
  background: #f9fafb;
}

/* Drop Zone Content */
.content {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 16px;
  max-width: 400px;
}

.icon {
  width: 48px;
  height: 48px;
  color: #9ca3af;
  transition: color 0.3s ease;
}

.dropzone:hover .icon,
.dropzone.dragActive .icon {
  color: #3b82f6;
}

.title {
  font-size: 18px;
  font-weight: 600;
  color: #111827;
  margin: 0;
}

.description {
  font-size: 14px;
  color: #6b7280;
  margin: 0;
  line-height: 1.5;
}

.button {
  display: inline-flex;
  align-items: center;
  gap: 8px;
  padding: 10px 20px;
  background: #3b82f6;
  color: white;
  border: none;
  border-radius: 8px;
  font-size: 14px;
  font-weight: 500;
  cursor: pointer;
  transition: all 0.2s ease;
}

.button:hover {
  background: #2563eb;
  transform: translateY(-1px);
  box-shadow: 0 4px 12px rgba(59, 130, 246, 0.3);
}

.button:active {
  transform: translateY(0);
}

.hiddenInput {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border: 0;
}

/* File List */
.filesList {
  margin-top: 24px;
}

.filesHeader {
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 16px;
}

.filesTitle {
  font-size: 16px;
  font-weight: 600;
  color: #111827;
  margin: 0;
}

.clearButton {
  background: none;
  border: none;
  color: #ef4444;
  font-size: 14px;
  cursor: pointer;
  padding: 4px 8px;
  border-radius: 4px;
  transition: background-color 0.2s ease;
}

.clearButton:hover {
  background: rgba(239, 68, 68, 0.1);
}

.fileItem {
  display: flex;
  align-items: center;
  gap: 12px;
  padding: 12px;
  border: 1px solid #e5e7eb;
  border-radius: 8px;
  background: white;
  margin-bottom: 8px;
  transition: all 0.2s ease;
}

.fileItem:hover {
  border-color: #d1d5db;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.05);
}

.filePreview {
  width: 40px;
  height: 40px;
  border-radius: 6px;
  overflow: hidden;
  flex-shrink: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background: #f3f4f6;
}

.filePreview img {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.fileIcon {
  width: 24px;
  height: 24px;
  color: #6b7280;
}

.fileInfo {
  flex: 1;
  min-width: 0;
}

.fileName {
  font-size: 14px;
  font-weight: 500;
  color: #111827;
  margin: 0 0 4px 0;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.fileDetails {
  display: flex;
  align-items: center;
  gap: 8px;
  font-size: 12px;
  color: #6b7280;
}

.fileSize {
  margin: 0;
}

.fileStatus {
  display: flex;
  align-items: center;
  gap: 4px;
  padding: 2px 6px;
  border-radius: 12px;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  letter-spacing: 0.025em;
}

.fileStatus.pending {
  background: #f3f4f6;
  color: #6b7280;
}

.fileStatus.uploading {
  background: #dbeafe;
  color: #1d4ed8;
}

.fileStatus.uploaded {
  background: #dcfce7;
  color: #16a34a;
}

.fileStatus.processing {
  background: #fef3c7;
  color: #d97706;
}

.fileStatus.completed {
  background: #dcfce7;
  color: #16a34a;
}

.fileStatus.failed {
  background: #fee2e2;
  color: #dc2626;
}

.fileActions {
  display: flex;
  align-items: center;
  gap: 8px;
}

.actionButton {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  border: none;
  border-radius: 6px;
  cursor: pointer;
  transition: all 0.2s ease;
}

.retryButton {
  background: #fef3c7;
  color: #d97706;
}

.retryButton:hover {
  background: #fcd34d;
  color: #92400e;
}

.removeButton {
  background: #fee2e2;
  color: #dc2626;
}

.removeButton:hover {
  background: #fca5a5;
  color: #b91c1c;
}

.progressBar {
  width: 100%;
  height: 4px;
  background: #f3f4f6;
  border-radius: 2px;
  overflow: hidden;
  margin-top: 8px;
}

.progressFill {
  height: 100%;
  background: linear-gradient(90deg, #3b82f6 0%, #1d4ed8 100%);
  border-radius: 2px;
  transition: width 0.3s ease;
}

.progressFill.completed {
  background: linear-gradient(90deg, #16a34a 0%, #15803d 100%);
}

.progressFill.failed {
  background: linear-gradient(90deg, #dc2626 0%, #b91c1c 100%);
}

/* Error Messages */
.errorList {
  margin-top: 16px;
}

.errorItem {
  display: flex;
  align-items: flex-start;
  gap: 8px;
  padding: 12px;
  background: #fef2f2;
  border: 1px solid #fecaca;
  border-radius: 8px;
  margin-bottom: 8px;
}

.errorIcon {
  width: 16px;
  height: 16px;
  color: #dc2626;
  flex-shrink: 0;
  margin-top: 2px;
}

.errorMessage {
  font-size: 14px;
  color: #b91c1c;
  margin: 0;
}

/* Responsive Design */
@media (max-width: 640px) {
  .dropzone {
    min-height: 160px;
    padding: 20px 16px;
  }

  .content {
    gap: 12px;
  }

  .icon {
    width: 40px;
    height: 40px;
  }

  .title {
    font-size: 16px;
  }

  .description {
    font-size: 13px;
  }

  .button {
    padding: 8px 16px;
    font-size: 13px;
  }

  .fileItem {
    padding: 10px;
    gap: 10px;
  }

  .filePreview {
    width: 36px;
    height: 36px;
  }

  .fileName {
    font-size: 13px;
  }

  .fileDetails {
    font-size: 11px;
  }
}

/* Animation for drag state */
@keyframes pulse {
  0%, 100% {
    opacity: 1;
  }
  50% {
    opacity: 0.8;
  }
}

.dropzone.dragActive .icon {
  animation: pulse 2s infinite;
}