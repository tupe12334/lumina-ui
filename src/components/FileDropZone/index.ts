export { FileDropZone } from './FileDropZone'
export type { FileDropZoneProps, FileDropZoneRef } from './FileDropZone'
export type { UseFileDropZoneConfig } from './application/types'