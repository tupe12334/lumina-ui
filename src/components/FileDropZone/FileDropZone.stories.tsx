import type { Meta, StoryObj } from '@storybook/react'
import { FileDropZone } from './FileDropZone'

const meta: Meta<typeof FileDropZone> = {
  title: 'Components/FileDropZone',
  component: FileDropZone,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: 'A modern file upload component with drag-and-drop support, file validation, and progress tracking.',
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    maxFiles: {
      control: { type: 'number', min: 1, max: 20 },
      description: 'Maximum number of files allowed',
    },
    maxFileSize: {
      control: { type: 'number' },
      description: 'Maximum file size in bytes',
    },
    acceptedTypes: {
      control: { type: 'object' },
      description: 'Array of accepted file types',
    },
    allowMultiple: {
      control: 'boolean',
      description: 'Whether multiple files can be selected',
    },
    disabled: {
      control: 'boolean',
      description: 'Whether the component is disabled',
    },
    hideFilesList: {
      control: 'boolean',
      description: 'Whether to hide the files list',
    },
    title: {
      control: 'text',
      description: 'Title text for the drop zone',
    },
    description: {
      control: 'text',
      description: 'Description text for the drop zone',
    },
    buttonText: {
      control: 'text',
      description: 'Text for the browse button',
    },
    onFilesAdded: {
      action: 'files added',
      description: 'Callback when files are added',
    },
    onFileRemoved: {
      action: 'file removed',
      description: 'Callback when a file is removed',
    },
    onError: {
      action: 'error',
      description: 'Callback when validation errors occur',
    },
  },
}

export default meta
type Story = StoryObj<typeof meta>

// Basic file drop zone
export const Default: Story = {
  args: {
    title: 'Upload Files',
    description: 'Drag and drop your files here, or click to browse',
    buttonText: 'Browse Files',
    maxFiles: 5,
    maxFileSize: 10 * 1024 * 1024, // 10MB
    acceptedTypes: ['*/*'],
    allowMultiple: true,
  },
}

// Single file upload
export const SingleFile: Story = {
  args: {
    title: 'Upload Document',
    description: 'Select a single document to upload',
    buttonText: 'Choose File',
    maxFiles: 1,
    allowMultiple: false,
    maxFileSize: 5 * 1024 * 1024, // 5MB
    acceptedTypes: ['application/pdf', 'application/msword', 'application/vnd.openxmlformats-officedocument.wordprocessingml.document'],
  },
}

// Image upload only
export const ImagesOnly: Story = {
  args: {
    title: 'Upload Images',
    description: 'Drag and drop your images here, or click to browse',
    buttonText: 'Select Images',
    maxFiles: 10,
    maxFileSize: 5 * 1024 * 1024, // 5MB
    acceptedTypes: ['image/*'],
    allowMultiple: true,
  },
}

// PDF documents only
export const PdfOnly: Story = {
  args: {
    title: 'Upload PDF Documents',
    description: 'Only PDF files are accepted',
    buttonText: 'Choose PDFs',
    maxFiles: 3,
    maxFileSize: 20 * 1024 * 1024, // 20MB
    acceptedTypes: ['application/pdf'],
    allowMultiple: true,
  },
}

// Disabled state
export const Disabled: Story = {
  args: {
    title: 'Upload Disabled',
    description: 'File upload is currently disabled',
    buttonText: 'Browse Files',
    disabled: true,
    maxFiles: 5,
    maxFileSize: 10 * 1024 * 1024,
  },
}

// Hidden files list
export const HiddenFilesList: Story = {
  args: {
    title: 'Upload Files',
    description: 'Files list is hidden for this example',
    buttonText: 'Browse Files',
    hideFilesList: true,
    maxFiles: 5,
    maxFileSize: 10 * 1024 * 1024,
  },
}

// Large file support
export const LargeFiles: Story = {
  args: {
    title: 'Upload Large Files',
    description: 'Upload files up to 100MB in size',
    buttonText: 'Select Large Files',
    maxFiles: 2,
    maxFileSize: 100 * 1024 * 1024, // 100MB
    acceptedTypes: ['*/*'],
    allowMultiple: true,
  },
}

// Custom styling
export const CustomStyling: Story = {
  args: {
    title: 'Custom Styled Upload',
    description: 'This drop zone has custom styling applied',
    buttonText: 'Upload Now',
    maxFiles: 5,
    maxFileSize: 10 * 1024 * 1024,
    className: 'custom-dropzone',
  },
  parameters: {
    docs: {
      description: {
        story: 'Example with custom CSS class applied for styling customization.',
      },
    },
  },
}

// Interactive example with callbacks
export const Interactive: Story = {
  args: {
    title: 'Interactive Upload',
    description: 'Check the Actions panel to see callback events',
    buttonText: 'Browse Files',
    maxFiles: 5,
    maxFileSize: 10 * 1024 * 1024,
    onFilesAdded: (files) => {
      console.log('Files added:', files)
    },
    onFileRemoved: (fileId) => {
      console.log('File removed:', fileId)
    },
    onError: (errors) => {
      console.log('Validation errors:', errors)
    },
  },
  parameters: {
    docs: {
      description: {
        story: 'Interactive example that logs events to the console. Open your browser developer tools to see the callback outputs.',
      },
    },
  },
}

// Small file limit for testing validation
export const ValidationDemo: Story = {
  args: {
    title: 'Validation Demo',
    description: 'Try uploading more than 2 files or files larger than 1MB',
    buttonText: 'Test Validation',
    maxFiles: 2,
    maxFileSize: 1024 * 1024, // 1MB
    acceptedTypes: ['image/*', 'application/pdf'],
    allowMultiple: true,
  },
  parameters: {
    docs: {
      description: {
        story: 'Demonstrates file validation with strict limits. Try uploading files that exceed the limits to see validation in action.',
      },
    },
  },
}