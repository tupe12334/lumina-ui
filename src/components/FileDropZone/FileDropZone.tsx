import { forwardRef } from 'react'
import { useFileDropZoneComponent } from './hooks/useFileDropZoneComponent'
import { FilesList } from './components/FilesList'
import { DropZoneArea } from './components/DropZoneArea'
import { type UseFileDropZoneConfig } from './application/types'
import styles from './FileDropZone.module.css'

export interface FileDropZoneProps extends UseFileDropZoneConfig {
  title?: string
  description?: string
  buttonText?: string
  icon?: string
  disabled?: boolean
  hideFilesList?: boolean
  className?: string
}

export interface FileDropZoneRef {
  removeFile: (fileId: string) => void
  retryFile: (fileId: string) => void
  clearAll: () => void
  openFileDialog: () => void
}

export const FileDropZone = forwardRef<FileDropZoneRef, FileDropZoneProps>(
  (props, ref) => {
    const {
      title = 'Upload Files',
      description = 'Drag and drop your files here, or click to browse',
      buttonText = 'Browse Files',
      icon,
      disabled = false,
      hideFilesList = false,
      className,
      ...config
    } = props

    const {
      files,
      isDragActive,
      handleDragEnter,
      handleDragOver,
      handleDragLeave,
      handleDrop,
      removeFile,
      retryFile,
      clearAll,
      getFileInputProps,
      fileInputRef,
      openFileDialog,
      maxFiles,
      maxFileSize
    } = useFileDropZoneComponent({ disabled, ...config }, ref)

    return (
      <div className={styles.container}>
        <DropZoneArea
          icon={icon}
          title={title}
          description={description}
          buttonText={buttonText}
          disabled={disabled}
          isDragActive={isDragActive}
          maxFiles={maxFiles}
          maxFileSize={maxFileSize}
          className={className}
          onDragEnter={handleDragEnter}
          onDragOver={handleDragOver}
          onDragLeave={handleDragLeave}
          onDrop={handleDrop}
          onClick={openFileDialog}
          onKeyDown={(e) => {
            if (!disabled && (e.key === 'Enter' || e.key === ' ')) {
              e.preventDefault()
              openFileDialog()
            }
          }}
          onButtonClick={(e) => { e.stopPropagation(); openFileDialog() }}
        >
          <input
            {...getFileInputProps()}
            ref={fileInputRef}
            className={styles.hiddenInput}
            disabled={disabled}
            aria-hidden="true"
          />
        </DropZoneArea>

        {!hideFilesList && files.length > 0 && (
          <FilesList files={files} onRemoveFile={removeFile} onRetryFile={retryFile} onClearAll={clearAll} showClearAll={files.length > 1} />
        )}
      </div>
    )
  }
)

FileDropZone.displayName = 'FileDropZone'