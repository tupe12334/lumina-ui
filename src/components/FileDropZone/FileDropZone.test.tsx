import { render, screen } from '@testing-library/react'
import { FileDropZone } from './FileDropZone'

describe('FileDropZone', () => {
  test('renders correctly with default props', () => {
    const { container } = render(<FileDropZone />)
    expect(container.firstChild).toMatchSnapshot()
  })

  test('renders with custom title and description', () => {
    render(
      <FileDropZone
        title="Custom Upload"
        description="Upload your custom files here"
      />
    )

    expect(screen.getByText('Custom Upload')).toBeInTheDocument()
    expect(screen.getByText('Upload your custom files here')).toBeInTheDocument()
  })

  test('renders with custom button text', () => {
    render(<FileDropZone buttonText="Select Files" />)
    expect(screen.getByText('Select Files')).toBeInTheDocument()
  })

  test('shows file limits in description', () => {
    render(
      <FileDropZone
        maxFiles={3}
        maxFileSize={5 * 1024 * 1024} // 5MB
      />
    )

    expect(screen.getByText(/Max 3 files/)).toBeInTheDocument()
    expect(screen.getByText(/Max 5MB per file/)).toBeInTheDocument()
  })

  test('renders in disabled state', () => {
    render(<FileDropZone disabled={true} />)

    const dropzone = screen.getByRole('button', { name: /file upload area/i })
    expect(dropzone).toHaveAttribute('aria-disabled', 'true')

    const button = screen.getByText('Browse Files')
    expect(button).toBeDisabled()
  })

  test('hides files list when hideFilesList is true', () => {
    const { container } = render(<FileDropZone hideFilesList={true} />)

    // FilesList should not be rendered
    expect(container.querySelector('[role="list"]')).not.toBeInTheDocument()
  })

  test('applies custom className', () => {
    const { container } = render(<FileDropZone className="custom-class" />)

    const dropzone = container.firstChild
    expect(dropzone).toHaveClass('custom-class')
  })

  test('renders with accessibility attributes', () => {
    render(<FileDropZone />)

    const dropzone = screen.getByRole('button', { name: /file upload area/i })
    expect(dropzone).toHaveAttribute('tabIndex', '0')
    expect(dropzone).toHaveAttribute('aria-label', 'File upload area')
    expect(dropzone).toHaveAttribute('aria-disabled', 'false')

    const fileInput = screen.getByLabelText(/browse files/i)
    expect(fileInput).toHaveAttribute('type', 'file')
    expect(fileInput).toHaveAttribute('aria-hidden', 'true')
  })

  test('configures file input based on props', () => {
    render(
      <FileDropZone
        allowMultiple={false}
        acceptedTypes={['image/*', 'application/pdf']}
      />
    )

    const fileInput = screen.getByLabelText(/browse files/i)
    expect(fileInput).not.toHaveAttribute('multiple')
    expect(fileInput).toHaveAttribute('accept', 'image/*,application/pdf')
  })

  test('allows multiple files by default', () => {
    render(<FileDropZone />)

    const fileInput = screen.getByLabelText(/browse files/i)
    expect(fileInput).toHaveAttribute('multiple')
  })
})