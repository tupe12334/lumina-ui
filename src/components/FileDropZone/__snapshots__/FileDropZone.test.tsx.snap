// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileDropZone > renders correctly with default props 1`] = `
<div
  class="_container_3af3de"
>
  <div
    aria-disabled="false"
    aria-label="File upload area"
    class="_dropzone_3af3de"
    role="button"
    tabindex="0"
  >
    <div
      class="_content_3af3de"
    >
      <svg
        class="_icon_3af3de"
        fill="none"
        height="48"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="48"
      >
        <path
          class=""
          d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
        />
        <polyline
          class=""
          points="7,10 12,5 17,10"
        />
        <line
          class=""
          x1="12"
          x2="12"
          y1="5"
          y2="15"
        />
      </svg>
      <h3
        class="_title_3af3de"
      >
        Upload Files
      </h3>
      <p
        class="_description_3af3de"
      >
        Drag and drop your files here, or click to browse
      </p>
      <button
        aria-label="Browse files"
        class="_button_3af3de"
        type="button"
      >
        Browse Files
      </button>
      <p
        class="_description_3af3de"
      >
        Max 
        5
         files â€¢ Max 
        10
        MB per file
      </p>
    </div>
    <input
      accept="*/*"
      aria-hidden="true"
      class="_hiddenInput_3af3de"
      multiple=""
      type="file"
    />
  </div>
</div>
`;
